tfimm/layers/lora.py

```diff
@@ -0,0 +1,92 @@
+import tensorflow as tf
+from dataclasses import dataclass
+from tfimm.architectures import convnext
+
+
+class LoraDense(tf.keras.layers.Dense):
+    def __init__(self, *args, lora_rank: int = 4, lora_alpha: float = 1, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.lora_rank = lora_rank
+        self.lora_alpha = lora_alpha
+        self.scaling = lora_alpha / lora_rank
+
+    def build(self, input_shape):
+        super().build(input_shape)
+        last_dim = self.kernel.shape[0]
+        self.kernel_0 = self.add_weight(
+            "kernel",
+            shape=[last_dim, self.units],
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+            dtype=self.dtype,
+            trainable=False,
+        )
+        self.kernel_lora_a = self.add_weight(
+            "kernel_lora_a",
+            shape=[last_dim, self.lora_rank],
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+            dtype=self.dtype,
+            trainable=True,
+        )
+        self.kernel_lora_b = self.add_weight(
+            "kernel_lora_b",
+            shape=[self.lora_rank, self.units],
+            initializer=tf.keras.initializers.Zeros(),
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+            dtype=self.dtype,
+            trainable=True,
+        )
+
+        self.kernel = (
+            self.kernel_0 + self.kernel_lora_a @ self.kernel_lora_b * self.scaling
+        )
```

Comment:
For completeness sake, we should also implement the `get_config()` method, so the layer can be serialized easily. We don't need it as such, but since it's not too complicated, let's be nice to the keras ecosystem.

Reply:
done